hw/ip/rom_ctrl/doc/_index.md

```diff
@@ -0,0 +1,285 @@
+This ROM has scrambled contents (scrambled with a fixed key, derived from a global constant).
+The controller is responsible for descrambling these contents on memory fetches.
+
+Unlike the [SRAM controller]({{< relref "hw/ip/sram_ctrl/doc" >}}), which performs the equivalent task for SRAM, the ROM controller also contains a *ROM checker* block.
+This ROM checker is used to compute a cryptographic hash of the ROM contents just after boot, detecting any malicious changes that have been made to the mask ROM when the system was at rest.
+
+## Features
+
+- Logic for memory and address descrambling
+- Post-boot ROM integrity check
+- Alert trigger and status CSRs for ROM integrity errors or FSM glitches.
+
+# Theory of Operations
+
+## Block Diagram
+
+The image below shows a high-level block diagram of the module.
+Blue boxes are instantiations of generic primitives that are used elsewhere on the chip.
+Green boxes are simple operations; the meat of the design is in the grey boxes.
+
+The upper half of the diagram shows paths for ROM reads when the system is in normal operation.
+The lower half of the diagram shows the ROM checker.
+This is triggered by the power manager early in the chip boot sequence to check validity of the ROM image.
+It runs exactly once, and releases the green multiplexer when it is done.
+
+![ROM Controller Block Diagram](rom_ctrl_blockdiag.svg)
+
+## ROM access when chip is in operation
+
+Once the chip has booted, ROM accesses are requested over the system TL-UL bus.
+These come in through the TL-UL SRAM adapter (top-left of block diagram).
+In normal operation, the green multiplexer will give access to these TL reads.
+The address is scrambled at the first substitution-permutation network (marked S&P in the diagram).
+
+In parallel with the ROM access, a reduced `prim_prince` primitive (5 rounds with latency 1; equivalent to the cipher used for SRAM) computes a 39-bit truncated keystream for the block.
+On the following cycle, the scrambled data from ROM goes through a substitution-permutation network and is then XOR'd with the keystream.
+This scheme is the same as that used by the [SRAM controller]({{< relref "hw/ip/sram_ctrl/doc" >}}), but is much simplified because the ROM doesn't have to deal with writes, byte accesses or key changes.
+
+The output from the XOR is the unscrambled 32-bit data, plus seven ECC bits.
+This finally gets passed through ECC decode and back to the TL-UL SRAM adapter.
+Any ECC decode error causes an `rerror` code to be reported by the TL access, a CSR to be set and a fatal alert to be generated.
+
+<div class="bd-callout bd-callout-warning">
+  <h5>ECC note</h5>
+
+It is likely that the system will move to augmenting the main TL-UL bus with ECC.
+If this happens, the design will change slightly, passing `rdata + ECC` directly from the descrambling XOR to the SRAM adapter.
+At this point, the ECC decode block and its associated CSRs can be removed.
```

Comment:
@tjaychen are we already at a point where we can say this is actually the case?

Reply:
Hah! I wrote that a while ago :-) I'll get rid of this now.